libp2p-swarm/src/swarm.rs

```diff
@@ -152,27 +151,24 @@ impl<T, C> SwarmController<T, C>
 }
 
 /// Future that must be driven to completion in order for the swarm to work.
-pub struct SwarmFuture<T, C, H, F>
+pub struct SwarmFuture<T, H, F>
     where T: MuxedTransport + 'static,      // TODO: 'static :-/
-          C: ConnectionUpgrade<T::RawConn> + 'static,      // TODO: 'static :-/
 {
-    upgraded: UpgradedNode<T, C>,
+    transport: T,
     handler: H,
-    new_listeners: mpsc::UnboundedReceiver<Box<Stream<Item = (Box<Future<Item = C::Output, Error = IoError>>, Multiaddr), Error = IoError>>>,
-    next_incoming: Box<Future<Item = (C::Output, Multiaddr), Error = IoError>>,
-    listeners: Vec<Box<Stream<Item = (Box<Future<Item = C::Output, Error = IoError>>, Multiaddr), Error = IoError>>>,
-    listeners_upgrade: Vec<(Box<Future<Item = C::Output, Error = IoError>>, Multiaddr)>,
-    dialers: Vec<(Box<Future<Item = C::Output, Error = IoError>>, Multiaddr)>,
-    new_dialers: mpsc::UnboundedReceiver<(Box<Future<Item = C::Output, Error = IoError>>, Multiaddr)>,
+    new_listeners: mpsc::UnboundedReceiver<T::Listener>,
+    next_incoming: T::Incoming,
+    listeners: Vec<T::Listener>,
+    listeners_upgrade: Vec<(T::ListenerUpgrade, Multiaddr)>,
+    dialers: Vec<(<T::Dial as IntoFuture>::Future, Multiaddr)>,
+    new_dialers: mpsc::UnboundedReceiver<(T::Dial, Multiaddr)>,
     to_process: Vec<future::Either<F, Box<Future<Item = (), Error = IoError>>>>,
     new_toprocess: mpsc::UnboundedReceiver<Box<Future<Item = (), Error = IoError>>>,
 }
 
-impl<T, C, H, If, F> Future for SwarmFuture<T, C, H, F>
+impl<T, H, If, F> Future for SwarmFuture<T, H, F>
```

Comment:
The `F` here is confusing and unnecessary, just use `If::Future` in its place (I'd call `If` `Fut`, too, because it's easy to confuse it with "if"). It makes it seem like this is more complicated than it really is. Also, can you give `H` a more useful name, like `Handler`?

Reply:
I can't really remove the `F`. The `H` is a closure that produces something that is turned into a `F`. The `SwarmFuture` has `H` and `F` as parameters because it stores them.

Reply:
```rust
impl<Trans, Hndlr, IntoFut> Future for SwarmFuture<Trans, Handler, IntoFut::Future>
where
    Trans: MuxedTransport + Clone + 'static,
    Handler: FnMut(Trans::Output, Multiaddr) -> IntoFut,
    IntoFut: IntoFuture<Item = (), Error = IoError>,
    IntoFut::Future: Future<Item = (), Error = IoError>, // Is this necessary given the above bound?
```